### DISCUSSION TITLE
Is hashFiles of actions/cache can be dynamic?

### DISCUSSION THREAD

**Discussion by joel** [community/community#26612] [Actions]
<p>I want to test against multiple rails versions, but it would be nice to benefit of the caching.</p>
<p>For instance I use:</p>
<pre><code class="lang-auto">- name: Install Ruby dependencies
      env:
        MATRIX_RAILS_VERSION: ${{ matrix.rails-version }}
      run: |
        export BUNDLE_GEMFILE="${GITHUB_WORKSPACE}/gemfiles/rails_${MATRIX_RAILS_VERSION}.gemfile"
</code></pre>
<p>But it doesn’t seem possible to passe dynamic reference of a file into <code>hashFiles</code></p>
<pre><code class="lang-auto">key: ${{ runner.os }}-${{ matrix.ruby-version }}-${{ matrix.rails-version }}-test-${{ hashFiles("${GITHUB_WORKSPACE}/gemfiles/rails_${{ matrix.rails-version }}.gemfile.lock") }}
</code></pre>
<p>Obviously this would have to exclude some versions like <code>head</code> and <code>master</code>.</p>

**Comment by Simran-B** [comment, marked as answer]
<p>Nesting <code>${{ }}</code> expressions inside of expressions is not supported. I’m also pretty sure that <code>$GITHUB_WORKSPACE</code> won’t work here because the value of <code>key</code> isn’t executed in a shell. You can try the following however:</p>
<p><code>key: ${{ runner.os }}-${{ matrix.ruby-version }}-${{ matrix.rails-version }}-test-${{ hashFiles(format('{0}/gemfiles/rails_{1}.gemfile.lock', github.workspace, matrix.rails-version)) }}</code></p>

**Comment by joel** [comment]
<p>Hi <a class="mention" href="/simran-b">@Simran-B</a> thanks for the reply, it’s much appreciated!</p>
<p>Indeed your proposal create properly the expected key:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/6/f/6f666ec78dc8723fccd70a20f71eb0f2b3017519.jpeg" data-download-href="/uploads/short-url/fTusxxbOE6pgQYQvZ5y0nXXQ5n3.jpeg?dl=1" title="Screen Shot 2021-03-31 at 9.09.45 PM" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/5789/181098679-e483609d-26a6-4c17-a8e2-f55a38c52006.jpeg" alt="Screen Shot 2021-03-31 at 9.09.45 PM" data-base62-sha1="fTusxxbOE6pgQYQvZ5y0nXXQ5n3" width="690" height="213" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/6/f/6f666ec78dc8723fccd70a20f71eb0f2b3017519_2_690x213.jpeg, https://aws1.discourse-cdn.com/github/optimized/3X/6/f/6f666ec78dc8723fccd70a20f71eb0f2b3017519_2_1035x319.jpeg 1.5x, https://aws1.discourse-cdn.com/github/optimized/3X/6/f/6f666ec78dc8723fccd70a20f71eb0f2b3017519_2_1380x426.jpeg 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/6/f/6f666ec78dc8723fccd70a20f71eb0f2b3017519_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screen Shot 2021-03-31 at 9.09.45 PM</span><span class="informations">1934×598 103 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>However, I have no clue why but the cache is never hits even if the tests are green and the key generated have exactly same value.</p>
<p>Anyway, the encouraged way to cache:</p>
<pre><code class="lang-auto">jobs:
  test:
    env:
      BUNDLE_GEMFILE: "${{ github.workspace }}/gemfiles/rails_${{ matrix.rails-version }}.gemfile"
    steps:
    - name: Set up Ruby ${{ matrix.ruby-version }}
      uses: ruby/setup-ruby@v1
      with:
        ruby-version: ${{ matrix.ruby-version }}
        bundler-cache: true
</code></pre>
<p>Generate a correct unique key and works!</p>
<pre><code class="lang-auto">Cache key: setup-ruby-bundler-cache-v3-ubuntu-20.04-ruby-3.0.0-/home/runner/work/swl/swl/gemfiles/rails_6.1.3.1.gemfile.lock-211b34df2618f8c21c0e5de2f6f46650fe73e6cba559b547f6fb289d6ed2cb4b
</code></pre>
<pre><code class="lang-auto">Cache key: setup-ruby-bundler-cache-v3-ubuntu-20.04-ruby-debug-revision-2a02b61fae2c5dcfaf123f43c08c7c7949c1790c-/home/runner/work/swl/swl/gemfiles/rails_main.gemfile.lock-4d904ceb814ad73ded167c4cd1d4035caadc4f9814ca9056d85a1daf598bc7a1
</code></pre>

**Comment by Simran-B** [comment]
<aside class="quote no-group" data-username="joel" data-post="3" data-topic="171342">
<div class="title">
<div class="quote-controls"></div>
 joel:</div>
<blockquote>
<p>the cache is never hits even if the tests are green and the key generated have exactly same value.</p>
</blockquote>
</aside>
<p>If it’s the exact same value, then none of the hashed files can have changed… Very peculiar. I wonder why it doesn’t work with the cache action. But good that you found an alternative solution!</p>

**Comment by joel** [comment]
<p>Yes indeed, it might have an edge case over there but I have not time to investigate further. you can have look though:</p>
<p>Here I added back the Manual Cache Setup</p><aside class="onebox githubfolder">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/joel/swl/runs/2269910877?check_suite_focus=true" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:500/500;"><img src="https://user-images.githubusercontent.com/5789/181098683-84059281-4f78-4aab-a821-aec3a7ab8349.png" class="thumbnail" width="500" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_500x500.png, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_750x750.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_1000x1000.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_10x10.png"></div>

<h3><a href="https://github.com/joel/swl/runs/2269910877?check_suite_focus=true" target="_blank" rel="noopener nofollow ugc">Build software better, together</a></h3>

<p><a href="https://github.com/joel/swl/runs/2269910877?check_suite_focus=true" target="_blank" rel="noopener nofollow ugc">swl/runs/2269910877</a></p>

  <p><span class="label1">GitHub is where people build software. More than 56 million people use GitHub to discover, fork, and contribute to over 100 million projects.</span></p>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>I did a nonsignificant change, pushed again:</p><aside class="onebox githubfolder">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/joel/swl/runs/2269940124?check_suite_focus=true" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:500/500;"><img src="https://user-images.githubusercontent.com/5789/181098684-a13952a8-b43f-43b0-a7ec-a4a5e6274938.png" class="thumbnail" width="500" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_500x500.png, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_750x750.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_1000x1000.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_10x10.png"></div>

<h3><a href="https://github.com/joel/swl/runs/2269940124?check_suite_focus=true" target="_blank" rel="noopener nofollow ugc">Build software better, together</a></h3>

<p><a href="https://github.com/joel/swl/runs/2269940124?check_suite_focus=true" target="_blank" rel="noopener nofollow ugc">swl/runs/2269940124</a></p>

  <p><span class="label1">GitHub is where people build software. More than 56 million people use GitHub to discover, fork, and contribute to over 100 million projects.</span></p>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>Same Key in both, but the key still not found.</p>
<pre><code class="lang-auto">Linux-3.0.0-6.1.3.1-test-7c88e8a6370e87e67677b3d3c8f7ae315acdc4f241f95167a042013df4231f17
</code></pre>
<p>Thanks anyway.</p>

**Comment by Simran-B** [comment]
<p>Shouldn’t it be <code>gemfiles/vendor/bundle</code> instead of <code>vendor/bundle</code> for the cache action <code>path</code>?</p>
